grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -67,13 +104,107 @@ def _UpdateProgress(
+
+  def _FileCollectionFromColumns(
+      self,
+      responses: Iterable[rdf_osquery.OsqueryResult],
+  ) -> None:
+    pathspecs = self._GetPathSpecsToCollect(responses)
+
+    stub = server_stubs.GetFileStat
+    for pathspec in pathspecs:
+      request = rdf_client_action.GetFileStatRequest(pathspec=pathspec)
+      self.CallClient(
+          stub,
+          request,
+          next_state=compatibility.GetName(self._StatForFileArrived))
+
+  def _StatForFileArrived(
+      self,
+      responses: flow_responses.Responses[rdf_client_fs.StatEntry],
+  ) -> None:
+    if not responses.success:
+      # TODO(simstoykov): Take note of this and report to the user with other
+      # flow results, instead of failing the flow.
+      raise flow_base.FlowError(
+        f"Error when attempted to stat file: {responses.status}")
+
+    if len(responses) != 1:
+      raise ValueError(
+          f"Response from stat has length {len(responses)}, instead of 1.")
+
+    stat_entry = list(responses)[0]
+
+    if stat_entry.st_size > FILE_COLLECTION_MAX_SINGLE_FILE_BYTES:
+      # TODO(simstoykov): Report to the user with other flow results instead of
+      # failing the flow.
+      file_path = stat_entry.pathspec.CollapsePath()
+      raise flow_base.FlowError(
+          f"File with path {file_path} is too big: {stat_entry.st_size} bytes "
+          f"when the limit is {FILE_COLLECTION_MAX_SINGLE_FILE_BYTES} bytes")
+
+    next_total_size = self.state.total_collected_bytes + stat_entry.st_size
+    if next_total_size > FILE_COLLECTION_MAX_TOTAL_BYTES:
+      # TODO(simstoykov): Consider reporting to the user and giving the
+      # collected files so far and the other results.
+      raise flow_base.FlowError("Files for collection exceed the total size "
+          f"limit of ${FILE_COLLECTION_MAX_SINGLE_FILE_BYTES} bytes")
+
+    self.StartFileFetch(stat_entry.pathspec)
+
+  def _GenerateTargetArchivePath(self, pathspec: rdf_paths.PathSpec) -> str:
+    path_info = rdf_objects.PathInfo.FromPathSpec(pathspec)
+    client_string_path = "/".join(path_info.components)
+
+    target_path = f"osquery_collected_files/{client_string_path}"
+
+    times_used_before = self.state.path_to_count.get(target_path, 0)
+    self.state.path_to_count[target_path] = times_used_before + 1
+    if times_used_before > 0:
+      target_path = f"{target_path}-{times_used_before}"
+
+    return target_path
+
   def Start(self):
-    super(OsqueryFlow, self).Start()
+    super(OsqueryFlow, self).Start(
+        file_size=FILE_COLLECTION_MAX_SINGLE_FILE_BYTES)
+
+    if len(self.args.file_collect_columns) > FILE_COLLECTION_MAX_COLUMNS:
+      message = (f"Requested file collection for "
+          f"{len(self.args.file_collect_columns)} columns, but the limit is"
+          f"{FILE_COLLECTION_MAX_COLUMNS} columns")
+      raise FileCollectionLimitsExceeded(message)
+
     self.state.progress = rdf_osquery.OsqueryProgress()
+    self.state.path_to_count = {}
```

Comment:
I don't think you need to put path_to_count into the state. It's only used in _GenerateTargetArchivePath and _GenerateTargetArchivePath is only used in GetFilesArchiveMappings. A cleaner solution would be to simply pass the path_to_count as an argument to _GenerateTargetArchivePath. What do you think?

Reply:
Yes that makes sense to me too. I got confused that we need to serialize this object and pass it across "state" functions, but there is no need as we only use it in `GetFilesArchiveMappings`.

Reply:
Okay I actually decided to split this functionality into a separate class for generating unique paths. Even though the logic is not super complicated, it bundles the state and the method that act on it; also I got the chance to make the code cleaner (subjective opinion).

Reply:
Looks good! I like the approach. Even more, I'd consider making _UniquePathGenerator a general-purpose class, so that we can use it for other flows as well (not in this PR).